### PR TITLE
adding Upper/Lower Bound Filter

### PR DESCRIPTION
**Pull request by b-slim**
This Pr is Fix to https://github.com/druid-io/druid/issues/1879

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/filter/BetweenFilter.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            } else if (betweenDimFilter.isLowerStrict()){
+              return (input.compareTo(betweenDimFilter.getLower()) < 0) && (input.compareTo(betweenDimFilter.getUpper())
+                                                                            >= 0);
+            } else if(betweenDimFilter.isUpperStrict()) {
+              return (input.compareTo(betweenDimFilter.getLower()) <= 0) && (input.compareTo(betweenDimFilter.getUpper())
+                                                                            > 0);
+            }
+            return (input.compareTo(betweenDimFilter.getLower()) <= 0) && (input.compareTo(betweenDimFilter.getUpper())
+                                                                                 >= 0);
+          }
```
Comment (himanshug) [inline processing/src/main/java/io/druid/segment/filter/BetweenFilter.java]: this is probably going to have very weird results for the `age` use case e.g. this will say that age "100" is smaller than age "20" . I believe, you would need to at least support a flag saying whether the comparison should be done by converting string to numbers.

Comment (b-slim) [inline processing/src/main/java/io/druid/segment/filter/BetweenFilter.java]: @himanshug thanks, done

**Comment by drcrallen** [comment]
@himanshug / @b-slim how would you feel if the numeric condition just used the comparator from io.druid.query.topn.AlphaNumericTopNMetricSpec ?

**Comment by drcrallen** [comment]
and possibly make it "alphaNumeric" instead of just numeric?

**Comment by himanshug** [comment]
@drcrallen @b-slim SGTM

**Comment by b-slim** [comment]
@drcrallen thanks, please check it out

File: processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import com.google.inject.Key;
+import io.druid.guice.GuiceInjectors;
+import io.druid.guice.annotations.Json;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+public class BetweenDimFilterTests
```
Comment (drcrallen) [inline processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java]: Is it possible to test each of the return cases for isAlphaNumeric both true and false?

Comment (b-slim) [inline processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java]: @drcrallen done !

File: processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+{
+  public BetweenDimFilterTests(BetweenDimFilter betweenDimFilter) {this.betweenDimFilter = betweenDimFilter;}
+
+  private final BetweenDimFilter betweenDimFilter;
+
+  @Parameterized.Parameters
+  public static Iterable<Object[]>  constructorFeeder(){
+
+    return ImmutableList.of(new Object[]{new BetweenDimFilter("dimension", "12", "15", null, null, false)},
```
Comment (drcrallen) [inline processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java]: There should be 8 cases here

**Comment by drcrallen** [comment]
Is it possible to test the apply for the 8 cases?

**Comment by xvrl** [comment]
@b-slim would it make sense to add less than and greater than filters as well while we're at it?
Between could be expressed as a combination of those or have it's own implementation, depending on whether it's worth to optimize, but it seems a bit strange to have `between` but not `< <= > >=`

File: processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java (hunk 3 above)
Comment (b-slim) [inline processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java]: @drcrallen i can add all the combinations, but for the purpose knowledge  i don't get why we need all the 8 combinations ? do you have a corner case in mind ?

**Comment by b-slim** [comment]
@xvrl i see how this can be useful, but i think we can have `between`, `< ` `<=` `>`  and `>=` all together.
i do prefer to have `between` since it saves the amount of bytes sent and and deserialized.
so i guess we can start with this, then i will send another PR of the rest of the troops.

**Comment by himanshug** [comment]
@xvrl did you mean adding more filters or updating the api for same filter?
@b-slim @xvrl  can we handle all cases by calling it a "bound" filter and based on whether user specified one of both of "lower" and "upper", the filter should be able to do appropriate thing.
so if user specifies
"lower" only : then it becomes a lower bound filter
"upper" only: then it becomes a upper bound filter
"lower" and "upper" both: then it becomes between filter

File: processing/src/main/java/io/druid/query/filter/DimFilterCacheHelper.java [hunk 4]
```diff
@@ -35,6 +35,7 @@
   static final byte JAVASCRIPT_CACHE_ID = 0x7;
   static final byte SPATIAL_CACHE_ID = 0x8;
   static final byte IN_CACHE_ID = 0x9;
+  public static byte BETWEEN_CACHE_ID = 0x10;
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/filter/DimFilterCacheHelper.java]: `0xA`?

File: processing/src/test/java/io/druid/query/timeseries/TimeseriesQueryRunnerTest.java [hunk 5]
```diff
@@ -2120,4 +2121,83 @@ public void testTimeseriesWithTimeColumn()
 
     TestHelper.assertExpectedResults(expectedResults, actualResults);
   }
+
+  @Test
+  public void testTimeseriesWithBetweenFilter1()
```
Comment (drcrallen) [inline processing/src/test/java/io/druid/query/timeseries/TimeseriesQueryRunnerTest.java]: Formatting here looks.... off

Comment (b-slim) [inline processing/src/test/java/io/druid/query/timeseries/TimeseriesQueryRunnerTest.java]: i don't think so

File: processing/src/main/java/io/druid/query/filter/DimFilterCacheHelper.java (hunk 4 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/DimFilterCacheHelper.java]: done

**Comment by b-slim** [comment]
@himanshug  and @xvrl please checkout this.

File: docs/content/querying/filters.md [hunk 6]
```diff
@@ -151,6 +151,97 @@ The grammar for a IN filter is as follows:
 }
 ```
 
+### Bound filter
+
+Bound filter can be used to filter by a comparing dimension values to an upper value or/and a lower value.
```
Comment (himanshug) [inline docs/content/querying/filters.md]: "...by comparing dimension..."

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
+package io.druid.query.filter;
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: license?

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: extra line?

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+  {
+    return alphaNumeric;
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] dimensionBytes = StringUtils.toUtf8(this.getDimension());
+    byte[] lowerBytes = StringUtils.toUtf8("lower:" + this.getLower());
+    byte[] upperBytes = StringUtils.toUtf8("upper:" + this.getUpper());
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: not sure why is encoding "lower/upper" necessary, it will only increase size of the key. you could use something like..
```
byte[] lowerBytes = (lower == null) ? EMPTY_BYTE_ARR : StringUtils.toUtf8(lower);
```
also, you wouldn't need any of the if statements later.

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+    if (this.getLower() == null) {
+      ByteBuffer betweenCacheBuffer = ByteBuffer.allocate(
+          6
+          + dimensionBytes.length
+          + upperBytes.length
+      );
+      betweenCacheBuffer.put(DimFilterCacheHelper.BETWEEN_CACHE_ID)
+                        .put(dimensionBytes)
+                        .put((byte) 0xFF)
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: probably putting in the extra "0xFF" is also not required.

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+  {
+    byte[] dimensionBytes = StringUtils.toUtf8(this.getDimension());
+    byte[] lowerBytes = StringUtils.toUtf8("lower:" + this.getLower());
+    byte[] upperBytes = StringUtils.toUtf8("upper:" + this.getUpper());
+    byte lowerStrictByte = (this.isLowerStrict() == false) ? (byte) 0 : (byte) 1;
+    byte upperStrictByte = (this.isUpperStrict() == false) ? (byte) 0 : (byte) 1;
+    byte AlphaNumericByte = (this.isAlphaNumeric() == false) ? (byte) 0 : (byte) 1;
+
+    if (this.getLower() == null) {
+      ByteBuffer betweenCacheBuffer = ByteBuffer.allocate(
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: boundCacheBuffer

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    if (isAlphaNumeric() != that.isAlphaNumeric()) {
+      return false;
+    }
+    if (!getDimension().equals(that.getDimension())) {
+      return false;
+    }
+    if (!getUpper().equals(that.getUpper())) {
+      return false;
+    }
+    return getLower().equals(that.getLower());
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: this is asuuming that both upper and lower are non-null

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+
+  }
+
+  @Override
+  public int hashCode()
+  {
+    int result = getDimension().hashCode();
+    result = 31 * result + getUpper().hashCode();
+    result = 31 * result + getLower().hashCode();
+    result = 31 * result + (isLowerStrict() ? 1 : 0);
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: assuming both upper and lower are non-null

File: processing/src/main/java/io/druid/query/filter/DimFilterCacheHelper.java [hunk 14]
```diff
@@ -35,6 +35,7 @@
   static final byte JAVASCRIPT_CACHE_ID = 0x7;
   static final byte SPATIAL_CACHE_ID = 0x8;
   static final byte IN_CACHE_ID = 0x9;
+  public static byte BETWEEN_CACHE_ID = 0xA;
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/filter/DimFilterCacheHelper.java]: BOUND_CACHE_ID

File: processing/src/main/java/io/druid/segment/filter/BoundFilter.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
+package io.druid.segment.filter;
```
Comment (himanshug) [inline processing/src/main/java/io/druid/segment/filter/BoundFilter.java]: license hdr

File: processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
+package io.druid.query.filter;
```
Comment (himanshug) [inline processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java]: license hdr also need to rename class to BoundDimFilterTests

File: processing/src/main/java/io/druid/segment/filter/BoundFilter.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+            Comparator<String> comparator;
+            if (boundDimFilter.isAlphaNumeric()) {
+              comparator = new AlphaNumericTopNMetricSpec(null).getComparator(null, null);
+            } else {
+              comparator = new LexicographicTopNMetricSpec(null).getComparator(null, null);
+            }
+
+            int lowerComparing = comparator.compare(input, boundDimFilter.getLower());
+            int upperComparing = comparator.compare(input, boundDimFilter.getUpper());
+            boolean lowerIsNull = boundDimFilter.getLower() == null;
```
Comment (himanshug) [inline processing/src/main/java/io/druid/segment/filter/BoundFilter.java]: hoping that comparator.compare(..) don't throw NPE if second arg was null

File: processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+    return ImmutableList.of(new Object[]{new BoundDimFilter("dimension", "12", "15", null, null, null)},
+                            new Object[]{new BoundDimFilter("dimension", "12", "15", null, true, false)},
+                            new Object[]{new BoundDimFilter("dimension", "12", "15", null, null, true)},
+                            new Object[]{new BoundDimFilter("dimension", "12", "15", null, true, true)},
+                            new Object[]{new BoundDimFilter("dimension", "12", "15", true, null, null)},
+                            new Object[]{new BoundDimFilter("dimension", "12", "15", true, null, true)},
+                            new Object[]{new BoundDimFilter("dimension", "12", "15", true, true, true)},
+                            new Object[]{new BoundDimFilter("dimension", "12", "15", true, true, false)});
+  }
```
Comment (himanshug) [inline processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java]: can you add cases for lower/upper being null

File: processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+
+  @Test
+  public void testSerDesBetweenFilter() throws IOException
```
Comment (himanshug) [inline processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java]: Bound

File: processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+  {
+    Injector defaultInjector = GuiceInjectors.makeStartupInjector();
+    ObjectMapper mapper = defaultInjector.getInstance(Key.get(ObjectMapper.class, Json.class));
+    String serBetweenDimFilter = mapper.writeValueAsString(boundDimFilter);
```
Comment (himanshug) [inline processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java]: can you use  TestHelper.getObjectMapper() ?

File: processing/src/test/java/io/druid/query/timeseries/TimeseriesQueryRunnerTest.java (hunk 5 above)
Comment (himanshug) [inline processing/src/test/java/io/druid/query/timeseries/TimeseriesQueryRunnerTest.java]: BoundFilter

File: docs/content/querying/filters.md [hunk 21]
```diff
@@ -151,6 +151,97 @@ The grammar for a IN filter is as follows:
...
 
+### Bound filter
+
+Bound filter can be used to filter by a comparing dimension values to an upper value or/and a lower value. 
+By default Comparison is string based and **case sensitive**.
+To use numeric comparison you can set `alphaNumeric` to `true`.
+By default the bound filter is a not a strict inclusion `inputString <= upper && inputSting >= lower`.
+  
+Note that if `alphaNumeric` is true, the column values that are not convertibles to `int` numbers will be treated as out of range.
+
```
Comment (himanshug) [inline docs/content/querying/filters.md]: can you give some examples on what is considered smaller/greater for both cases?

File: docs/content/querying/filters.md (hunk 6 above)
Comment (b-slim) [inline docs/content/querying/filters.md]: done

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java (hunk 7 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: done

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java (hunk 8 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: done

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java (hunk 9 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: done

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java (hunk 10 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: we need this i guess

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java (hunk 11 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: done

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java (hunk 12 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: done

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java (hunk 13 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: done

File: processing/src/main/java/io/druid/query/filter/DimFilterCacheHelper.java (hunk 14 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/DimFilterCacheHelper.java]: done

File: processing/src/main/java/io/druid/segment/filter/BoundFilter.java (hunk 15 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/filter/BoundFilter.java]: done

File: processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java (hunk 16 above)
Comment (b-slim) [inline processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java]: done

File: processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java (hunk 20 above)
Comment (b-slim) [inline processing/src/test/java/io/druid/query/filter/BetweenDimFilterTests.java]: i think you mean TestUtils.getTestObjectMapper(), this will not work due to dependencies issues

File: docs/content/querying/filters.md (hunk 21 above)
Comment (b-slim) [inline docs/content/querying/filters.md]: i think it is obvious ? do you have any pitfalls you want to pinpoint ?

File: processing/src/main/java/io/druid/segment/filter/BoundFilter.java (hunk 17 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/filter/BoundFilter.java]: good one, please check it out

File: docs/content/querying/filters.md (hunk 21 above)
Comment (himanshug) [inline docs/content/querying/filters.md]: no pitfalls, but, not sure if person just reading documentation would understand what "out of range" meant in this context, so asked for examples, would be a nice to have.

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+                        .put(dimensionBytes)
+                        .put((byte) 0xFF)
+                        .put(lowerBytes)
+                        .put((byte) 0xFF)
+                        .put(lowerStrictByte)
+                        .put((byte) 0xFF)
+                        .put(AlphaNumericByte);
+      return betweenCacheBuffer.array();
+
+    } else {
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: after taking care of nulls to EMPTY bytes array, why do we need to handle the null cases for getUpper/Lower ?
what goes wrong we do not put 0xFF as separators ? (puting them for no use just increases the size of resulting byte[])

**Comment by b-slim** [comment]
@himanshug can u please check this

**Comment by fjy** [comment]
@b-slim Merge conflicts
@himanshug any more comments?

File: processing/src/main/java/io/druid/query/filter/BoundDimFilter.java [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+    boundCacheBuffer.put(DimFilterCacheHelper.BOUND_CACHE_ID)
+                    .put(boundType)
+                    .put(upperStrictByte)
+                    .put(lowerStrictByte)
+                    .put(AlphaNumericByte)
+                    .put((byte) 0xFF)
+                    .put(dimensionBytes)
+                    .put((byte) 0xFF)
+                    .put(upperBytes)
+                    .put((byte) 0xFF)
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/filter/BoundDimFilter.java]: I think another PR introduced a separate variable for 0xFF in this context, can you use that?

**Comment by himanshug** [comment]
:+1: after https://github.com/druid-io/druid/pull/1936#discussion_r47162619 is resolved, pls squash your commits.

**Comment by b-slim** [comment]
@himanshug done

**Comment by himanshug** [comment]
:+1: for me

**Comment by b-slim** [comment]
@fjy this is ready to go can you merge it please

**Comment by fjy** [comment]
I read this over again, and I think it is cool. :+1:
